// @novu/react/dist/esm/components/subscription/Subscription.js
// src/components/subscription/Subscription.tsx
import React, { useMemo } from "react";
import { useNovu } from "../../hooks/NovuProvider.js";
import { NovuUI } from "../NovuUI.js";
import { withRenderer } from "../Renderer.js";
import { DefaultSubscription } from "./DefaultSubscription.js";
import { jsx } from "react/jsx-runtime";
var SubscriptionInternal = withRenderer((props) => {
  const { container, localization, appearance, ...defaultSubscriptionProps } = props;
  const novu = useNovu();
  const options = useMemo(() => {
    return {
      container,
      localization,
      appearance,
      options: novu.options
    };
  }, [localization, appearance, container, novu.options]);
  if (isWithChildrenProps(props)) {
    const clonedChildren = React.Children.map(props.children, (child) => {
      if (React.isValidElement(child)) {
        return React.cloneElement(child, {
          ...child.props,
          topicKey: defaultSubscriptionProps.topicKey,
          identifier: defaultSubscriptionProps.identifier,
          preferences: defaultSubscriptionProps.preferences
        });
      }
      return child;
    });
    return /* @__PURE__ */ jsx(NovuUI, { options, novu, children: clonedChildren });
  }
  return /* @__PURE__ */ jsx(NovuUI, { options, novu, children: /* @__PURE__ */ jsx(DefaultSubscription, { ...defaultSubscriptionProps }) });
});
SubscriptionInternal.displayName = "SubscriptionInternal";
var Subscription = React.memo((props) => {
  return /* @__PURE__ */ jsx(SubscriptionInternal, { ...props });
});
function isWithChildrenProps(props) {
  return "children" in props;
}
export {
  Subscription
};
