// @novu/react/dist/cjs/components/subscription/Subscription.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/subscription/Subscription.tsx
var Subscription_exports = {};
__export(Subscription_exports, {
  Subscription: () => Subscription
});
module.exports = __toCommonJS(Subscription_exports);
var import_react = __toESM(require("react"), 1);
var import_NovuProvider = require("../../hooks/NovuProvider.cjs");
var import_NovuUI = require("../NovuUI.cjs");
var import_Renderer = require("../Renderer.cjs");
var import_DefaultSubscription = require("./DefaultSubscription.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var SubscriptionInternal = (0, import_Renderer.withRenderer)((props) => {
  const { container, localization, appearance, ...defaultSubscriptionProps } = props;
  const novu = (0, import_NovuProvider.useNovu)();
  const options = (0, import_react.useMemo)(() => {
    return {
      container,
      localization,
      appearance,
      options: novu.options
    };
  }, [localization, appearance, container, novu.options]);
  if (isWithChildrenProps(props)) {
    const clonedChildren = import_react.default.Children.map(props.children, (child) => {
      if (import_react.default.isValidElement(child)) {
        return import_react.default.cloneElement(child, {
          ...child.props,
          topicKey: defaultSubscriptionProps.topicKey,
          identifier: defaultSubscriptionProps.identifier,
          preferences: defaultSubscriptionProps.preferences
        });
      }
      return child;
    });
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NovuUI.NovuUI, { options, novu, children: clonedChildren });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NovuUI.NovuUI, { options, novu, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DefaultSubscription.DefaultSubscription, { ...defaultSubscriptionProps }) });
});
SubscriptionInternal.displayName = "SubscriptionInternal";
var Subscription = import_react.default.memo((props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SubscriptionInternal, { ...props });
});
function isWithChildrenProps(props) {
  return "children" in props;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Subscription
});
