// @novu/react/dist/esm/server/index.js
// src/server/index.tsx
import { ShadowRootDetector } from "../components/ShadowRootDetector.js";
import { PreferenceLevel, SeverityLevelEnum, WorkflowCriticalityEnum } from "@novu/js";
import { Fragment, jsx } from "react/jsx-runtime";
function Inbox(props) {
  return /* @__PURE__ */ jsx(ShadowRootDetector, {});
}
function InboxContent() {
}
function Notifications() {
}
function Preferences() {
}
function Bell() {
}
function NovuProvider(props) {
  return /* @__PURE__ */ jsx(Fragment, { children: props.children });
}
function Subscription() {
  return /* @__PURE__ */ jsx(ShadowRootDetector, {});
}
function SubscriptionButton() {
}
function SubscriptionPreferences() {
}
function useNovu() {
  return null;
}
function useCounts(_) {
  return {
    isLoading: false,
    isFetching: false,
    refetch: () => Promise.resolve()
  };
}
function useNotifications(_) {
  return {
    isLoading: false,
    isFetching: false,
    hasMore: false,
    readAll: () => Promise.resolve({ data: void 0, error: void 0 }),
    seenAll: () => Promise.resolve({ data: void 0, error: void 0 }),
    archiveAll: () => Promise.resolve({ data: void 0, error: void 0 }),
    archiveAllRead: () => Promise.resolve({ data: void 0, error: void 0 }),
    refetch: () => Promise.resolve(),
    fetchMore: () => Promise.resolve()
  };
}
function usePreferences(_) {
  return {
    isLoading: false,
    isFetching: false,
    refetch: () => Promise.resolve()
  };
}
function useSchedule(_) {
  return {
    isLoading: false,
    isFetching: false,
    refetch: () => Promise.resolve()
  };
}
function useSubscription(_) {
  return {
    isLoading: false,
    isFetching: false,
    refetch: () => Promise.resolve()
  };
}
function useCreateSubscription(_ = {}) {
  return {
    isCreating: false,
    error: void 0,
    create: () => Promise.resolve({ data: void 0, error: void 0 })
  };
}
function useUpdateSubscription(_ = {}) {
  return {
    isUpdating: false,
    error: void 0,
    update: () => Promise.resolve({ data: void 0, error: void 0 })
  };
}
function useRemoveSubscription(_ = {}) {
  return {
    isRemoving: false,
    error: void 0,
    remove: () => Promise.resolve({ data: void 0, error: void 0 })
  };
}
function useSubscriptions(_) {
  return {
    isLoading: false,
    isFetching: false,
    refetch: () => Promise.resolve()
  };
}
export {
  Bell,
  Inbox,
  InboxContent,
  Notifications,
  NovuProvider,
  PreferenceLevel,
  Preferences,
  SeverityLevelEnum,
  Subscription,
  SubscriptionButton,
  SubscriptionPreferences,
  WorkflowCriticalityEnum,
  useCounts,
  useCreateSubscription,
  useNotifications,
  useNovu,
  usePreferences,
  useRemoveSubscription,
  useSchedule,
  useSubscription,
  useSubscriptions,
  useUpdateSubscription
};
